NextJS - Failed to start Quarkus

### Describe the bug

When I create a fresh NextJS application with Quarkus, the DEV mode works without any problems.
The QuarkusTests also work, but I have the following errors with the QuarkusIntegrationTests:

`2024-01-03 14:37:07,674 ERROR [io.qua.run.Application] (main) Failed to start application (with profile [prod]): java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:111)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:44)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:124)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: 0
        at java.base/java.lang.StringLatin1.charAt(StringLatin1.java:48)
        at java.base/java.lang.String.charAt(String.java:1512)
        at io.vertx.ext.web.handler.impl.StaticHandlerImpl.setIndexPage(StaticHandlerImpl.java:635)
        at io.quarkiverse.quinoa.QuinoaUIResourceHandler.createStaticHandler(QuinoaUIResourceHandler.java:67)
        at io.quarkiverse.quinoa.QuinoaUIResourceHandler.<init>(QuinoaUIResourceHandler.java:37)
        at io.quarkiverse.quinoa.QuinoaRecorder.quinoaHandler(QuinoaRecorder.java:41)
        at io.quarkus.deployment.steps.QuinoaProcessor$runtimeInit397778252.deploy_0(Unknown Source)
        at io.quarkus.deployment.steps.QuinoaProcessor$runtimeInit397778252.deploy(Unknown Source)
        ... 13 more`

### Package.json:
`
{
  "name": "lcm-webui",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "react": "^18",
    "react-dom": "^18",
    "next": "14.0.4"
  },
  "devDependencies": {
    "typescript": "^5",
    "@types/node": "^20",
    "@types/react": "^18",
    "@types/react-dom": "^18",
    "autoprefixer": "^10.0.1",
    "postcss": "^8",
    "tailwindcss": "^3.3.0",
    "eslint": "^8",
    "eslint-config-next": "14.0.4"
  }
}
`

### next.config.js
`const nextConfig = {
    output: 'export',
}
module.exports = nextConfig
`

### Quinoa version
2.3.1

### Quarkus version
3.6.4

### Build / Runtime
Next.js

### Package Manager
NPM

--- a/deployment/src/main/java/io/quarkiverse/quinoa/deployment/ForwardedDevProcessor.java
+++ b/deployment/src/main/java/io/quarkiverse/quinoa/deployment/ForwardedDevProcessor.java
@@ -198,7 +198,7 @@ public class ForwardedDevProcessor {
                 return;
             }
             LOG.infof("Quinoa is forwarding unhandled requests to port: %d", devProxy.get().getPort());
-            final QuinoaHandlerConfig handlerConfig = toHandlerConfig(quinoaConfig, false, httpBuildTimeConfig);
+            final QuinoaHandlerConfig handlerConfig = toHandlerConfig(quinoaConfig, true, httpBuildTimeConfig);
             routes.produce(RouteBuildItem.builder().orderedRoute("/*", QUINOA_ROUTE_ORDER)
                     .handler(recorder.quinoaProxyDevHandler(handlerConfig, vertx.getVertx(), devProxy.get().getHost(),
                             devProxy.get().getPort(),

--- a/deployment/src/main/java/io/quarkiverse/quinoa/deployment/QuinoaProcessor.java
+++ b/deployment/src/main/java/io/quarkiverse/quinoa/deployment/QuinoaProcessor.java
@@ -259,7 +259,7 @@ public class QuinoaProcessor {
                 directory = uiResources.get().getDirectory().get().toAbsolutePath().toString();
             }
             final QuinoaHandlerConfig handlerConfig = toHandlerConfig(configuredQuinoa.resolvedConfig(),
-                    !launchMode.getLaunchMode().isDevOrTest(),
+                    launchMode.getLaunchMode() == LaunchMode.DEVELOPMENT,
                     httpBuildTimeConfig);
             resumeOn404.produce(new ResumeOn404BuildItem());
             routes.produce(RouteBuildItem.builder().orderedRoute("/*", QUINOA_ROUTE_ORDER)
@@ -357,7 +357,7 @@ public class QuinoaProcessor {
      * Ignored directories include any that start with DOT "." like ".next" or ".svelte", also "node_modules" and any
      * of the framework build directories.
      *
-     * @param filePath the file path to check
+     * @param relativeFilePath the file path to check
      * @return true if it is a directory that should be scanned for changes, false if it should be ignored
      */
     private static boolean shouldWatch(String relativeFilePath) {

--- a/deployment/src/main/java/io/quarkiverse/quinoa/deployment/config/QuinoaConfig.java
+++ b/deployment/src/main/java/io/quarkiverse/quinoa/deployment/config/QuinoaConfig.java
@@ -141,13 +141,20 @@ public interface QuinoaConfig {
         }).stream().map(s -> s.startsWith("/") ? s : "/" + s).collect(toList());
     }
 
-    static QuinoaHandlerConfig toHandlerConfig(QuinoaConfig config, boolean prodMode,
+    static QuinoaHandlerConfig toHandlerConfig(QuinoaConfig config, boolean devMode,
             final HttpBuildTimeConfig httpBuildTimeConfig) {
         final Set<String> compressMediaTypes = httpBuildTimeConfig.compressMediaTypes.map(Set::copyOf).orElse(Set.of());
-        final String indexPage = !isDevServerMode(config) ? config.indexPage()
-                : config.devServer().indexPage().orElse(config.indexPage());
-        return new QuinoaHandlerConfig(getNormalizedIgnoredPathPrefixes(config), indexPage, prodMode,
+        final String indexPage = resolveIndexPage(config, devMode);
+        return new QuinoaHandlerConfig(getNormalizedIgnoredPathPrefixes(config), indexPage, devMode,
                 httpBuildTimeConfig.enableCompression, compressMediaTypes, config.devServer().directForwarding());
+    }
+
+    private static String resolveIndexPage(QuinoaConfig config, boolean devMode) {
+        if (!devMode) {
+            // Make sure we never return the devServer.indexPage() in non-dev mode
+            return config.indexPage();
+        }
+        return isDevServerMode(config) ? config.devServer().indexPage().orElse(config.indexPage()) : config.indexPage();
     }
 
     private static Optional<String> readExternalConfigPath(Config config, String key) {

--- a/runtime/src/main/java/io/quarkiverse/quinoa/QuinoaHandlerConfig.java
+++ b/runtime/src/main/java/io/quarkiverse/quinoa/QuinoaHandlerConfig.java
@@ -9,18 +9,18 @@ import io.quarkus.runtime.annotations.RecordableConstructor;
 public class QuinoaHandlerConfig {
     public final List<String> ignoredPathPrefixes;
     public final String indexPage;
-    public final boolean prodMode;
+    public final boolean devMode;
     public final boolean enableCompression;
     public final Set<String> compressMediaTypes;
 
     public final boolean devServerDirectForwarding;
 
     @RecordableConstructor
-    public QuinoaHandlerConfig(List<String> ignoredPathPrefixes, String indexPage, boolean prodMode, boolean enableCompression,
+    public QuinoaHandlerConfig(List<String> ignoredPathPrefixes, String indexPage, boolean devMode, boolean enableCompression,
             Set<String> compressMediaTypes, boolean devServerDirectForwarding) {
         this.ignoredPathPrefixes = ignoredPathPrefixes;
         this.indexPage = "/".equals(indexPage) ? "" : indexPage;
-        this.prodMode = prodMode;
+        this.devMode = devMode;
         this.enableCompression = enableCompression;
         this.compressMediaTypes = compressMediaTypes;
         this.devServerDirectForwarding = devServerDirectForwarding;
@@ -33,7 +33,7 @@ public class QuinoaHandlerConfig {
         if (o == null || getClass() != o.getClass())
             return false;
         QuinoaHandlerConfig that = (QuinoaHandlerConfig) o;
-        return prodMode == that.prodMode && enableCompression == that.enableCompression
+        return devMode == that.devMode && enableCompression == that.enableCompression
                 && devServerDirectForwarding == that.devServerDirectForwarding
                 && Objects.equals(ignoredPathPrefixes, that.ignoredPathPrefixes) && Objects.equals(indexPage, that.indexPage)
                 && Objects.equals(compressMediaTypes, that.compressMediaTypes);
@@ -41,7 +41,7 @@ public class QuinoaHandlerConfig {
 
     @Override
     public int hashCode() {
-        return Objects.hash(ignoredPathPrefixes, indexPage, prodMode, enableCompression, compressMediaTypes,
+        return Objects.hash(ignoredPathPrefixes, indexPage, devMode, enableCompression, compressMediaTypes,
                 devServerDirectForwarding);
     }
 }

--- a/runtime/src/main/java/io/quarkiverse/quinoa/QuinoaUIResourceHandler.java
+++ b/runtime/src/main/java/io/quarkiverse/quinoa/QuinoaUIResourceHandler.java
@@ -70,7 +70,7 @@ class QuinoaUIResourceHandler implements Handler<RoutingContext> {
                 : StaticHandler.create(QuinoaRecorder.META_INF_WEB_UI);
         staticHandler.setDefaultContentEncoding(StandardCharsets.UTF_8.name());
         staticHandler.setIndexPage(config.indexPage);
-        staticHandler.setCachingEnabled(config.prodMode);
+        staticHandler.setCachingEnabled(!config.devMode);
         return staticHandler;
     }
 

